libraries/aws/shadow/src/aws_iot_shadow_api.c

```diff
@@ -519,57 +536,56 @@ static AwsIotShadowError_t _modifyCallbackSubscriptions( IotMqttConnection_t mqt
+                         pCallbackSuffixLength[ type ] );
+
+        IotLogDebug( "%s subscription for %.*s",
+                     mqttOperation == IotMqtt_SubscribeSync ? "Adding" : "Removing",
+                     operationTopicLength + pCallbackSuffixLength[ type ],
+                     pTopicFilter );
+
+        /* Set the members of the MQTT subscription. */
+        subscription.qos = IOT_MQTT_QOS_1;
+        subscription.pTopicFilter = pTopicFilter;
+        subscription.topicFilterLength = ( uint16_t ) ( operationTopicLength + pCallbackSuffixLength[ type ] );
+        subscription.callback.pCallbackContext = NULL;
+        subscription.callback.function = pCallbackWrapper[ type ];
+
+        /* Call the MQTT operation function. */
+        mqttStatus = mqttOperation( mqttConnection,
+                                    &subscription,
+                                    1,
+                                    0,
+                                    _AwsIotShadowMqttTimeoutMs );
+
+        /* Check the result of the MQTT operation. */
+        if( mqttStatus != IOT_MQTT_SUCCESS )
+        {
+            IotLogError( "Failed to %s callback for %.*s %s callback, error %s.",
+                         mqttOperation == IotMqtt_SubscribeSync ? "subscribe to" : "unsubscribe from",
+                         pSubscription->thingNameLength,
+                         pSubscription->pThingName,
+                         _pAwsIotShadowCallbackNames[ type ],
+                         IotMqtt_strerror( mqttStatus ) );
+
+            /* Convert the MQTT "NO MEMORY" error to a Shadow "NO MEMORY" error. */
+            status = SHADOW_CONVERT_STATUS_CODE_MQTT_TO_SHADOW( mqttStatus );
+        }
     }
 
-    /* Place the callback suffix in the topic filter. */
-    ( void ) memcpy( pTopicFilter + operationTopicLength,
-                     pCallbackSuffix[ type ],
-                     pCallbackSuffixLength[ type ] );
-
-    IotLogDebug( "%s subscription for %.*s",
-                 mqttOperation == IotMqtt_SubscribeSync ? "Adding" : "Removing",
-                 operationTopicLength + pCallbackSuffixLength[ type ],
-                 pTopicFilter );
-
-    /* Set the members of the MQTT subscription. */
-    subscription.qos = IOT_MQTT_QOS_1;
-    subscription.pTopicFilter = pTopicFilter;
-    subscription.topicFilterLength = ( uint16_t ) ( operationTopicLength + pCallbackSuffixLength[ type ] );
-    subscription.callback.pCallbackContext = NULL;
-    subscription.callback.function = pCallbackWrapper[ type ];
-
-    /* Call the MQTT operation function. */
-    mqttStatus = mqttOperation( mqttConnection,
-                                &subscription,
-                                1,
-                                0,
-                                _AwsIotShadowMqttTimeoutMs );
-
-    /* Check the result of the MQTT operation. */
-    if( mqttStatus != IOT_MQTT_SUCCESS )
+    if( status == IOT_MQTT_SUCCESS )
```

Comment:
This should be `AWS_IOT_SHADOW_SUCCESS`. Also, this `if` block could also have been an `else` following the `if` on 566

Reply:
yes